src/MessageSystem.cpp

```diff
@@ -0,0 +1,347 @@
+                break;
+            case MsgParamType::String:
+                WRITE_STRING(std::get<std::string>(param.data).c_str());
+                break;
+            case MsgParamType::Entity:
+                WRITE_ENTITY(std::get<int>(param.data));
+                break;
+        }
+    }
+
+    MESSAGE_END();
+}
+
+void Message::clearParams()
+{
+    m_params.clear();
+}
+
+int Message::getParamInt(std::size_t index) const
+{
+    return getParam<int>(index);
+}
+float Message::getParamFloat(std::size_t index) const
+{
+    return getParam<float>(index);
+}
+const char *Message::getParamString(std::size_t index) const
+{
+    return getParam<std::string>(index).data();
+}
+
+void Message::setParamInt(std::size_t index, int value)
+{
+    setParam(index, value);
+}
+void Message::setParamFloat(std::size_t index, float value)
+{
+    setParam(index, value);
+}
+void Message::setParamString(std::size_t index, const char *string)
+{
+    setParam(index, string);
+}
+
+void Message::addParam(MsgParamType type, std::variant<int, float, std::string> &&data)
+{
+    MessageParam param;
+    param.type = type;
+    param.data = data;
+
+    m_params.push_back(param);
+}
+
+void Message::setParam(std::size_t index, std::variant<int, float, std::string> &&data)
+{
+    m_params[index].data = data;
+}
+
+std::size_t Message::getParams() const
+{
+    return m_params.size();
+}
+
+MsgParamType Message::getParamType(std::size_t index) const
+{
+    return m_params[index].type;
+}
+
+int Message::getDest() const
+{
+    return m_dest;
+}
+
+int Message::getType() const
+{
+    return m_type;
+}
+
+const float *Message::getOrigin() const
+{
+    return m_origin;
+}
+
+IEdict *Message::getEdict() const
+{
+    return m_edict.get();
+}
+
+MessageHook::MessageHook(int msgType, MessageHandler handler, std::any cbData, bool post)
+    : m_msgType(msgType), m_handler(handler), m_cbData(cbData), m_post(post), m_active(true)
+{
+}
+
+// IMessageHook
+void MessageHook::enable()
+{
+    m_active = true;
+}
+void MessageHook::disable()
+{
+    m_active = false;
+}
+bool MessageHook::isActive() const
+{
+    return m_active;
+}
+
+int MessageHook::getMsgType() const
+{
+    return m_msgType;
+}
+
+// MessageHook
+bool MessageHook::getPost() const
+{
+    return m_post;
+}
+MessageHandler MessageHook::getHandler() const
+{
+    return m_handler;
+}
+std::any MessageHook::getCbData() const
+{
+    return m_cbData;
+}
+bool MessageHook::getActive() const
+{
+    return m_active;
+}
+
+MessageHook *MessageHooks::addHook(int msgType, MessageHandler handler, std::any cbData, bool post)
+{
+    return m_handlers.emplace_back(std::make_unique<MessageHook>(msgType, handler, cbData, post)).get();
+}
+void MessageHooks::removeHook(IMessageHook *hook)
+{
+    auto iter = m_handlers.begin();
+    while (iter != m_handlers.end())
+    {
+        if ((*iter).get() == hook)
+        {
+            m_handlers.erase(iter);
+            break;
+        }
+        ++iter;
+    }
```

Comment:
`std::remove_if` / `std::erase_if`?

Reply:
We could use it to make code more clean.

Reply:
Requires  lastest c++ standard but project is c++17 for now

Reply:
@Mistrick only `std::erase_if`. `std::remove_if` can still be used.